### ISSUE TITLE
Get FunctorFilter on par with the other MTL typeclasses

### ISSUE THREAD

**Issue by Baccata** [typelevel/cats#2556]
Provided a `FunctorFilter` instance for `F[_]` (and potentially additional constraints), it seems possible to provide `FunctorFilter` instances for `StateT`, `ReaderT`, etc ... and get it on par with the other MTL typeclasses. 

I was in need of such instances and quickly hacked together an implementation, without checking the laws, and without trying really hard to relax the constraint as much as I could, but the following seems to work : 

```scala

  implicit def stateTFunctorFilter[F[_], S](
      implicit F: Monad[F],
      FF: FunctorFilter[F]): FunctorFilter[StateT[F, S, ?]] =
    new FunctorFilter[StateT[F, S, ?]] {
      override def functor: Functor[StateT[F, S, ?]] = Functor[StateT[F, S, ?]]

      override def mapFilter[A, B](fa: StateT[F, S, A])(
          f: A => Option[B]): StateT[F, S, B] =
        fa.flatMapF(a => FF.mapFilter(F.pure(a))(f))
    }

  implicit def readerTFunctorFilter[F[_], S](
      implicit F: Monad[F],
      FF: FunctorFilter[F]): FunctorFilter[ReaderT[F, S, ?]] =
    new FunctorFilter[ReaderT[F, S, ?]] {
      override def functor: Functor[ReaderT[F, S, ?]] =
        Functor[ReaderT[F, S, ?]]

      override def mapFilter[A, B](fa: ReaderT[F, S, A])(
          f: A => Option[B]): ReaderT[F, S, B] =
        fa.flatMapF(a => FF.mapFilter(F.pure(a))(f))
    }
```

On another note, still related to FunctorFilter, it'd be great to have `FunctorFilter#fromOption`, and `OptionOps.LiftToPartiallyApplied#flatten` (shame that the `apply` there is already taken by ApplicativeError)

**Comment by barambani** [comment]
I have some time to have a look at this if that's ok

**Comment by barambani** [comment]
@Baccata , I opened an initial PR with the instances described here and I wanted to finish it with the last bits at the bottom of your comment. Could you share with me a sample type signature of what you would expect from `FunctorFilter#fromOption` and `OptionOps.LiftToPartiallyApplied#flatten` or may be just a use case ? I drafted something but I see some overlapping with the `FunctorFilter`'s syntax for `Option` so I'd like to make sure I got correctly what you mean. Thanks a lot.

**Comment by Baccata** [comment]
Writing from phone, so apologies in advance.

I've revised my opinion on this. Considering the Functor nature of the typeclass in question, having a "fromOption" on FunctorFilter would probably not be right since you'd probably need to rely on some applicative nature of the target type for implementation. 

I was initially also thinking that having a `def flatten(implicit F : Applicative[A], FF : FunctorFilter[A]) : F[A]` on the OptionOps.LiftToPartiallyApplied would be handy, but it'd be the exact same thing as calling pure then flatten, so no need for it.

In other words i think the PR in its current state is probably good enough.

**Comment by barambani** [comment]
👍 , thanks for the quick reply.

**Comment by kailuowang** [comment]
I believe this is closed by #2594, feel free to reopen if we have leftovers.
